### PR TITLE
Fixes #13302 - associate and create an operating system during registration

### PR DESCRIPTION
**Pull request by jlsherrill**

### CODE REVIEW

**Comment by jlsherrill** [comment]
Note this requires https://github.com/Katello/katello/pull/5713
but going ahead and opening.  This creates (if doesn't exist) the operating system and asociates it upon registration.  It does kind of suck that the rhsm facts for operating system are so different from puppet which is the gold standard in forman (assuming it comes from LSB), but its what we deal with.  Could possibly add more later on such as for scientific and oracle linux, but these are probably good to start with.

**Comment by ehelms** [comment]
This does a translation to ensure that it creates repositories the same way (same naming) that Puppet facts would?

**Comment by jlsherrill** [comment]
Operating systems i assume you mean (not repositories) and yes,  the goal is to create the same operating system as if a puppet checkin was happening.  That way if you are registering to install puppet and then puppet runs the OS doesn't change

**Comment by jlsherrill** [comment]
Curious if you had any thoughts @stbenjam

**Comment by stbenjam** [comment]
Sure, seems reasonable. I might consider having the Fact Parser assign the operating system here: https://github.com/Katello/katello/blob/master/app/models/katello/rhsm_fact_parser.rb, if possible.  Not sure if that works with the workflow you want, but if it does, Foreman will assign the host's OS to whatever the operatingsystem method returns.

**Comment by jlsherrill** [comment]
@stbenjam yeah i had originally gone down that path, but my concern was:

a) if we make a mistake the OS will constantly flip back and forth as puppet and rhsm checks in
b) the operating system shouldn't ever change after registration (if it does a new registration is expected as we've never really handled OS upgrades).

**Comment by stbenjam** [comment]
Yea, a) makes sense, although that's probably something we should address in Foreman, as we now have many fact providers (RHSM, Chef, Salt, Puppet) with parsers so you really could end up with all kinds of flipping back and forth.

Although b) isn't always true, as you can go from 6.5 -> 6.6 for example, right? Unless the subscription manager facts don't have the minor.

**Comment by jlsherrill** [comment]
@stbenjam doh! completely forgot about the minor version!  :cake: 

A user is unlikely to be running more than one of puppet, chef, and salt but they may use ansible and puppet so i agree it is a common problem. 

I could go either way and will only take a minute to switch, are you leaning either way @stbenjam ?

**Comment by stbenjam** [comment]
It's really hard to say, because users will be confused if they update a non-Puppet managed host from 7.1 to 7.2 and it never updates in the Host index.  If it's consistent with the names Foreman creates from Puppet, I'd go with the fact parser.

Oh, and another question, we create OS's whenever you sync a distribution - are *those* names consistent with how Puppet calls them?

**Comment by jlsherrill** [comment]
@stbenjam i  believe they are and that someone did some work to make sure they were, i'd have to double check.

i will change to the fact importer then

**Comment by jlsherrill** [comment]
updated to use fact parser/importer

File: app/models/katello/rhsm_fact_parser.rb [hunk 1]
```diff
@@ -36,10 +36,20 @@ def get_interfaces
...
+
+      os_name = ::Katello::Candlepin::Consumer.distribution_to_puppet_os(name)
+      if os_name
+        major, minor = version.split('.')
+        os_attributes = {:major => major, :minor => minor || '', :name => os_name}
+        ::Operatingsystem.find_by(os_attributes) || ::Operatingsystem.create!(os_attributes)
+      end
     end
 
+    #reqiured to be defined, even if they return nil
```
Comment (parthaa) [inline app/models/katello/rhsm_fact_parser.rb]: nit pick ```required```

**Comment by parthaa** [comment]
LGTM. Tested on rhel 7.1 client. OS created cleanly and the UI was consistent. ACK after changing that nitpick

**Pull request merged**